CHANGELOG.unreleased.md

```diff
@@ -338,7 +338,29 @@ const b = new (c()!)();
 const b = new c()!();
 
 // Output (Prettier master)
-const b = new (c()!)();
+const b = new (c())!();
+```
+
+### Javascript: Address parentheses bugs for edge cases with call and new. ([#6148] by [@bakkot])
+
+Adding all and only the necessary parentheses when mixing `new` with function calls is tricky. This change fixes two issues where necessary parentheses were omitted and one when redundant parentheses were added.
+
+<!-- prettier-ignore -->
+```ts
+// Input
+new (x()``.y)();
+new (x()!.y)();
+new e[f().x].y()
+
+// Output (Prettier stable)
+new x()``.y();
+new x()!.y();
+new e[(f()).x].y();
+
+// Output (Prettier master)
+new (x())``.y();
```

Comment:
Looks very ugly

Reply:
See #6147. I figured it was more important to get it right to start with.

Reply:
@evilebottnawi This will be fixed by #6412.